typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1015 @@
+  y: number;
+  0: string;
+}
+```
+
+Now, using mapped types, we will create a type that will pick only some of the properties.
+Word `extends` here says that generic `K` will contain a subset of union properties keys `T`.
+And the resulting type will contain only properties from `K`.
+
+```typescript
+type Pick<T, K extends keyof T> = {
+  [P in K]: T[P];
+};
+
+type PointValues = Pick<Point, "x" | 0> // type PointValues = { x: number; 0: number }
+```
+
+The next example will be more complex.
+We will try to create the type which produces a type with getters for all string properties.
+
+To implement it, let's first define a `Getter` helper type.
+Capitalize is a built-in type, which may be applied to the string type.
+So, for `T`, which extends string, we can produce getter specific name.
+
+```typescript
+type Getter<T extends string> = `get${Capitalize<T>}`;
+```
+
+And finally, we can implement the `Getters` type.
+Type `Extract` extracts from a type all union members that are assignable to some type.
+So, here for each property name `K` in `T` that is assignable to `string`, we create a getter property name with type `() => T[K]`.
+
+```typescript
+type Getters<T> = {
+    [K in keyof T as Getter<Extract<K, string>>]: () => T[K];
+}
+
+type PointGetters = Getters<Point> // type PointValues = { getX: () => number; getY: () => number; }
+```
+
+#### Conditional types
+
+Another great feature is conditional types.
+It allows you to choose the type based on some condition expressed as another type.
+With `extends`, you are able to make a condition: `T extends U ? X : Y`.
+It also allows you to choose not only from two but from many types: `T extends U ? X : T extends W ? Y : Z`.
+With such functionality, you can implement things that resemble [type families](https://serokell.io/blog/type-families-haskell), although they are not quite like them.
+Anyway, in conjunction with mapped types, it is a great feature to manage your types.
+
+One thing to note is union types.
+Conditional types together with union types are called distributive conditional types, and how they work might be a little bit confusing.
+But we will explain them with examples.
+
+Here is a basic example of a conditional type.
+Type `B` extends `A`, but type `C` doesn't.
+And `Condition` type will return one type if `T extends V` and another one otherwise.
+
+```typescript
+type A = { x: number };
+type B = { x: number; y: number };
+type C = { y: number };
+
+type Condition<T, V> = T extends V ? number : string;
+
+type N = Condition<B, A>; // type N = number
+type E = Condition<C, A>; // type E = string
+```
+
+Let's move to distributive conditional types.
+`Exclude` type may look strange, but what if we will pass union types here?
+It will simply exclude such types in union `T` which exists in union `U` setting `never` in other cases, where the [`never`](https://www.tutorialsteacher.com/typescript/typescript-never) type describes the type of values that will never occur.
```

Comment:
I think there is a typo here, it will return `never` exactly when it matches, according to code, no?

Reply:
Oh, not in other cases) \
Basically it is not returning smth but, 🤔, setting the type \
But `_ ? _ : _` construction is known as a ternary operator and saying `return`, as for me, sounds more understandable here \
So, I will change it, but maybe we can leave `return` here